BREAKING Overlay transaction support.

cheme opened the pull request:
BREAKING: this pr adds externalities, and change mutability of existing one, the corresponding 

This PR is a refactor of overlay_change to allow transactional support, it is similar to #2980 goals, but does not stack `Ext`, it only stacks storage.

It also switches from a stack of hashmap (previously `prospective` and `top`) to a single hashmap (containing history of values) and a transaction global state. Those values with state are managed
with 'historied-data' crate (simple vec of data to query in front of a reference global state).

Under this design access to data is not badly impacted by the number of open transactional layers.
I did fuzz this code a bit against a partial simple layered hashmap implementation.

Usage from a runtime with a function, in a similar way as `ext_try` from #2980, there is the very simple `with_transaction` function: internally it uses three host functions `ext_start_transaction`, `ext_commit_transaction` and `ext_discard_transaction`. This does not look as good as the single `ext_try` but is clearer: memory mgmt seems way simplier (as there is none).

Note that to call global state action, modification of local values need to be synchronize.
eg discard_transaction on states must be follow by apply_discard_transaction for all values related to this state and then follow by ensure_running.

Technically we only maintain a counter of current number of stacked transaction as a global state (this start at 1 and can be 0 only to update some values: case of discarding content, but will then return to 1 through 'finalize_discard' call).
Local state is either committed or the number of stacked transaction when the value was changed, this state is stored with the value history.

polkadot companion: https://github.com/paritytech/polkadot/pull/999

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -624,7 +624,22 @@ pub mod ext {
+		// Transactional Context
+		//================================
+
+		/// Open an new transactional context.
```
```suggestion
		/// Open a new transactional context.
```

pepyakin commented on core/sr-io/without_std.rs:
```diff
@@ -624,7 +624,22 @@ pub mod ext {
+		/// Open an new transactional context.
+		fn ext_start_transaction();
+
+		/// Close and dissmiss current transactional context.
```
```suggestion
		/// Close and dismiss current transactional context.
```

pepyakin replied:
also, let's stick to the `layer` term?

pepyakin commented on core/state-machine/src/basic.rs:
```diff
@@ -172,6 +172,19 @@ impl<H: Hasher> Externalities<H> for BasicExternalities where H::Out: Ord {
+	}
+
+	fn discard_transaction(&mut self) {
+		panic!("No support for transaction");
```
Can you explain why these are noops/panics?

cheme replied:
BasicExternalities is currently only use for genesis init, so this makes failing transaction panic during genesis build. That is not ideal, especially considering BasicExteranalities could be use in different context.
So I see two way of doing things:

- document basic externalities limitation properly (no transaction, no offchain ext) and document genesis initialisation limitation.
- implement transaction by switching back to internal overlay, this is quite straight forward. May be a bit heavy, but considering that it is use for genesis only it may not be such an issue.

cc/ @thiolliere since I know you know this code.

thiolliere replied:
I don't see any other use than genesis for now.

I'm also mixed: providing such features for building genesis storage is of no interest. but having an API that have panicking function because you're not suppose to use it is also not very good.

Also I see some assert in the building of the genesis in srml-staking, maybe no panic policy is not same importance for genesis build ?

thiolliere replied:
maybe it is better to have no-op and a warning like it does for offchain workers

DemiMarie-parity replied:
> Also I see some assert in the building of the genesis in srml-staking, maybe no panic policy is not same importance for genesis build ?

I would assume so.  No untrusted inputs go into building genesis, and it is much better to catch problems before a chain is launched.

pepyakin commented on core/state-machine/src/ext.rs:
```diff
@@ -343,30 +348,38 @@ mod tests {
 				digest_interval: 0,
 				digest_levels: 0,
 			}),
+			changes:OverlayedChangeSet {
```
```suggestion
			changes: OverlayedChangeSet {
```

pepyakin commented on core/state-machine/Cargo.toml:
```diff
@@ -19,3 +19,13 @@ num-traits = "0.2"
+
+[features]
+default = ["bench"]
+bench = []
```
Can you leave a comment on this feature? Why would one enable this?

cheme replied:
yes was for bench, I cannot find a good way to activate a function for bench only (here 'bench' as a default feature is clearly wrong). Since it was only to make a function public, I am simply switching the function to public for now ('set_storage').

pepyakin commented on srml/support/src/storage/mod.rs:
```diff
@@ -27,6 +27,24 @@ pub mod storage_items;
 pub mod unhashed;
 pub mod hashed;
 
+/// Execute under a transactional layer.
```
```suggestion
/// Execute under a transactional layer.
///
```

pepyakin replied:
(if you make put space after the first line, rustdoc makes the first line a summary. )

pepyakin commented on srml/support/src/storage/mod.rs:
```diff
@@ -27,6 +27,24 @@ pub mod storage_items;
+/// If the result of execution is an error,
+/// the transactional layer get reverted; otherwhise
+/// it is committed.
+pub fn with_transaction<R>(f: impl FnOnce() -> Result<R, &'static str>) -> Result<R, &'static str> {
```
```suggestion
pub fn with_transaction<R, E>(f: impl FnOnce() -> Result<R, E>) -> Result<R, E> {
```

thiolliere commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -48,22 +67,333 @@ pub struct OverlayedValue {
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub(crate) struct History<V>(usize, [Option<(V, usize)>; ALLOCATED_HISTORY], Vec<(V, usize)>);
```
why not named fields ?

thiolliere commented:
> Currently transactional primitives methods name can be a bit confusing (start_transaction do open an new transactional layer but next to this method there is also submit_transaction that submit an actual chain transaction).
So another name than 'transaction' should be suitable.

Maybe start_transactional_layer

cheme commented:
> Maybe start_transactional_layer
I forgot to update my comment, I change it to 'storage_start_transaction' but maybe 'start_transactional_layer' is better?

pepyakin commented:
I am actually not sure about the wording "layer". I.e. transactional layer of what?

DemiMarie-parity commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -74,17 +408,158 @@ impl OverlayedChangeSet {
+	}
+
+	/// Iterator over current state of the overlay.
+	pub fn top_iter_overlay(&self) -> impl Iterator<Item = (&Vec<u8>, &OverlayedValue)> {
```
Should return `&[u8]`, to add room for future changes.

DemiMarie-parity commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -74,17 +408,158 @@ impl OverlayedChangeSet {
+	}
+
+	/// Iterator over current state of the overlay.
+	pub fn top_iter(&self) -> impl Iterator<Item = (&Vec<u8>, Option<&Vec<u8>>)> {
```
Same

cheme commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -48,22 +67,337 @@ pub struct OverlayedValue {
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub(crate) struct History<V> {
```
I did switch from Vec looking where perf was lost, and drop this quick implementation without thinking to much, but indeed smallvec do things a bit more efficiently here (got a 5% on bench but the bench is not super relevant) and more importantly less code is better, thanks :)

cheme commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -74,17 +408,158 @@ impl OverlayedChangeSet {
+		let mut i = self.history.len();
+		while i > 0 {
+			i -= 1;
+			match self.history[i] {
```
This would break the property of this cache. I will write a gc method for cleaning (get_mut does only clean the simple case without going to deep as we can expect many get_mut call between dropped).
I tend to agree that memory gc could be triggered on 'drop_perspective' and 'commit_perspectve' : those operation being not frequent.
Gc on those operation could also be conditional to actual used (estimated) memory: we may want to chain small extrinsic operation without memory mgmt.

Truncating history can also be done up to the last commit after a full gc (we will need a offset).

pepyakin commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -48,22 +68,307 @@ pub struct OverlayedValue {
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub(crate) struct History<V> (SmallVec<[(V, usize); ALLOCATED_HISTORY]>);
```
```suggestion
pub(crate) struct History<V>(SmallVec<[(V, usize); ALLOCATED_HISTORY]>);
```

pepyakin commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -48,22 +68,307 @@ pub struct OverlayedValue {
+///
+/// First field is the current size of the history.
+/// Second field are statically allocated values for performance purpose.
+/// Third field are dynamically allocated values.
```
I am not sure what these comments refer to?

Can we introduce a nominal type with explicit fields?

cheme replied:
My mistake, I miss the documentation update when switching to smallvec.

DemiMarie-parity commented:
@cheme LGTM once you resolve the merge conflicts.

cheme commented:
I tried to ran some additional benches against implementation with multiple hashmap (only for set storage and get storage in root storage):
(branch https://github.com/cheme/substrate/tree/overlay_change_b1 to ..._b5).
Measuring: https://gist.github.com/cheme/5440ab82bf612f1db326025c488a6cd9
I also tried using a modified hashbrown which only hash once the key for all hashmap overlays and whereas is an ok improvment over running multiple hashmap, I do not feel like it is preferable to current pr.

cheme commented:
The possibility to get a top iter from child_iter function is only here to allow same iterator types for both layer (needed in change trie), but it is far from ideal. I kept the top_iter variant only because it produces a simpler iterator types, I am not sure it is worth keeping the function.

pepyakin commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -104,61 +306,68 @@ impl OverlayedChanges {
 	/// `None` can be used to delete a value specified by the given key.
-	pub(crate) fn set_storage(&mut self, key: Vec<u8>, val: Option<Vec<u8>>) {
+	pub fn set_storage(&mut self, key: Vec<u8>, value: Option<Vec<u8>>) {
+		self.operation_from_last_gc += DEFAULT_GC_CONF.operation_cost(value.as_ref());
```
What is the reason of having a dedicated structure if we access it as a global constant anyway? Should it be actually a field on `self`?

cheme replied:
I started to write the code parameterize-able so I kept the structure for if/when it needs to become parameterize-able.

cheme commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,488 @@
+
+use crate::State as TransactionState;
+use rstd::vec::Vec;
+use rstd::vec;
```
that's the vec macro, I need it for no_std.

cheme commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,488 @@
+	/// This method shall only be call after a `get_mut` where
+	/// the returned index indicate that a `set` will result
+	/// in appending a value.
+	pub fn unsafe_push(&mut self, value: V, history_index: usize) {
```
clearly a better name, I will try to remember it.

cheme commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,488 @@
+	/// Debugging function for test and fuzzing.
+	pub fn internal_item_counts(&self) -> usize {
+		self.0.len()
+	}
```
it is just here to expose `len` as public (mainly to test gc).

pepyakin replied:
yeah, but I mean look at the [`Self::len`](https://github.com/paritytech/substrate/pull/3263/files/b534b3d2a070ba7fd7c264ff042f4a20b0ba92bd#diff-e6eb5324fa37219c00c89cbfb41558bcR85), it has the same implementation basically. Why don't just call it instead of `internal_item_counts`.

cheme replied:
I can do like that https://github.com/paritytech/substrate/pull/3263/commits/d9d21185c012f14687d7a322fda1d9bced8af570 , but I like `internal_item_counts` name better.

DemiMarie-parity commented:
Would it be possible to use interior mutability so that `get` can also free memory?

cheme commented:
@DemiMarie-parity I did try it a bit: https://github.com/cheme/substrate/compare/overlay_change...cheme:overlay_change_inner_mut .
My conclusion is that it is rather impacting (see comment in state-machine) due to need to replace some reference by 'Ref'.
I could force it to return a & by using unchecked access, but then there will be no use to use inner mutability, but simply an unsafe mutability (I still wanted to compile the linked branch to assert if it may be possible, but when considering the expected gain I am not sure it is worth it).

bkchr commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -49,43 +54,246 @@ pub struct OverlayedValue {
+		self.top.retain(|_, h_value| h_value.get_mut_pruning(history, eager()).is_some());
+		self.children.retain(|_, m| {
+			m.retain(|_, h_value| h_value.get_mut_pruning(history, eager()).is_some());
+			m.len() > 0
```
```suggestion
			!m.is_empty()
```
?

cheme commented on .gitlab-ci.yml:
```diff
@@ -102,9 +102,12 @@ cargo-check-benches:
   <<:                              *docker-env
   script:
     - BUILD_DUMMY_WASM_BINARY=1 time cargo +nightly check --benches --all
+    - cd ./core/state-machine/fuzz
```
The fuzz generated cargo.toml isolate it from a workspace explicitly, I am not sure if there is any good reason for it, "cargo test" behaving oddly on the project can be one.

cheme commented on core/primitives/src/traits.rs:
```diff
@@ -178,6 +178,16 @@ pub trait Externalities<H: Hasher> {
+	/// Create a new transactional layer.
+	fn storage_start_transaction(&mut self);
+
+	/// Discard a transactional layer, pending changes of every trasaction below this layer are
```
Depends on how you visualize stacking transaction, I guess it is more natural to stack up indeed. I will use 'behind' as wording, seems neutral.

cheme commented on core/state-machine/fuzz/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[dependencies.libfuzzer-sys]
+git = "https://github.com/rust-fuzz/libfuzzer-sys.git"
+
+# Prevent this from interfering with workspaces
```
This one is generated from cargo fuzz, I think some unexpected behavior (I did only observe one with cargo test).

cheme commented on core/state-machine/src/overlayed_changes.rs:
```diff
@@ -49,43 +54,246 @@ pub struct OverlayedValue {
+	}
+
+	/// Iterator over current state of a given overlay, values only.
+	pub fn iter_values(
```
I also prefer using '&Vec<u8>' than '&[u8]' when inner storage is a vec and there is no reason to slice it.
But on previous pr an probably this one too, that is a recurring thing that I had to change, I guess there is a consensus that api should not use '&Vec<u8>', main argument being probably having a more future proof api, which can be a point for a trait; in this case it is not a trait so I am be very tempted to follow your suggestion. But I see it as being argued against in the future (it is still public api).

kianenigma replied:
(even in the current PR you had it first as `Vec<_>` and then changed to slice as a reaction to a review)

cheme commented:
> I'm not entirely sure we need all this complexity? Do we really need nested transactions?

at this point it seems like something many want, for instance contract are managing their own change deltas when it is really not something complicated to have in our overlay (we already have two layer so adding unbound number of layer does not add a lot of complexity, when it seems to be a rather big complication for ink).

> After reading all this, I think just having a HashMap that takes all values in the transaction context and committing them afterwards, would have made this much easier. While the current approach of having a history is nice and cool, we have stuff like garbage collection?

Yes stacking hashmap is what is done usually, there is some advantages:
- dropping a layer can free memory at once by dropping the hashmap
- commiting a layer can be done by merging hashmap which is probably a bit more efficient than the gc. But you can also choose to not merge or not gc.
- simplicity: my fuzzing is done over such implementation.

but there is also some point that I do not really enjoy:
- accessing a value means iterating over all hashmap until a match, for instance a get on a missing value will go through n hashmap query. So overlays got direct impact on perf.
On the overhand with this pr implementation number of overlays do not impact, that is the number of time a same value get change in overlay that impacts (and in the consideration of n change value array access).
Still when overlay got dropped there is a slight advantages to hashmap stacked (but I do not see it as relevent for get value).
So clearly the interesting thing of this implementation is that you do not have to care on how many overlay are in use.
- multiple hash as currently implemented for committed and pending layer, does not merge map and usually return iterator over pending chained with committed iterator so committed value is the one that prevails, that is not the case with this pr: iterator only return valid value.

For memory usage, I cannot really say, I think if there is a lot of stacked values stack hashmap will take more memory, otherwhise it may be the oposite. Considering at one side the overhead of a vec per key and at the other side the overhead of a hashmap key per transaction layer.  


>Do you have made any benchmarks?

I firstly bench on the no overlay case: it was running worst, then replacing the initial vec by a smallvec obviously bring things on par (or slightly better I do not remember).

More recently https://github.com/paritytech/substrate/pull/3263#issuecomment-528340805 (hidden amongst all conversation), do not really change my perspective (the variant 'memoization' is running against a modified hashbrown crate where we memoize result of hashing key (so it does not hash key on every layers)).

cheme commented:
I found the previous bench rather unsatisfactory, during lunch I was thinking it only resume to comparing hasmap of vector with vector of hasmap, and got a quick doubt:
The first case (this pr) intuitively seem rather bad because the number of structure grows with the number of key when for the other case it only grows with the number of overlay. But comparing a hashmap with a vec does not make sense.
So I added a insertion only test (where this pr may be bad) with different number of keys and got some ok result:
- https://github.com/cheme/substrate/tree/overlay_change_b1 (this pr): https://gist.github.com/cheme/d82ca9413419b2a466d422c5acfcbe37
- https://github.com/cheme/substrate/tree/overlay_change_b2 (quick impl of set_storage with stacked hashmap) :  https://gist.github.com/cheme/698bc96b0758b671c511161e49e1fa20

things seems better for stacked hashmap only when there is a single hashmap (the size 2 small vec certainly explain that), but all in all it seems are not that different (10 to 20% difference), and the bench is certainly not realistic.

bkchr commented on core/state-machine/Cargo.toml:
```diff
@@ -15,12 +15,23 @@ trie = { package = "substrate-trie", path = "../trie" }
 primitives = { package = "substrate-primitives", path = "../primitives" }
 panic-handler = { package = "substrate-panic-handler", path = "../panic-handler" }
 codec = { package = "parity-scale-codec", version = "1.0.0" }
+historied-data = { path = "../utils/historied-data", features = ["std"] }
```
```suggestion
historied-data = { path = "../utils/historied-data" }
```

bkchr commented on core/state-machine/Cargo.toml:
```diff
@@ -15,12 +15,23 @@ trie = { package = "substrate-trie", path = "../trie" }
 hex-literal = "0.2.1"
+criterion = "0.2"
+rand = { version = "0.7.1", features = ["small_rng"] }
+historied-data = { path = "../utils/historied-data", features = ["std", "test"] }
```
```suggestion
historied-data = { path = "../utils/historied-data", features = [ "test" ] }
```

bkchr commented on core/sr-api-macros/tests/runtime_calls.rs:
```diff
@@ -131,6 +130,23 @@ fn use_trie_function() {
 }
 
+#[test]
+fn use_history_data() {
```
Both of these tests should be live somewhere else.

bkchr commented on core/state-machine/Cargo.toml:
```diff
@@ -15,12 +15,23 @@ trie = { package = "substrate-trie", path = "../trie" }
 
 [features]
 default = []
+test = [
```
Should be deleted.

bkchr commented on core/state-machine/fuzz/src/lib.rs:
```diff
@@ -0,0 +1,254 @@
+		println!("fuzzing: \n {:x?}", (&actions, &values));
+		println!("input: \n {:?}", &input);
+	}
+
```
```suggestion
```

bkchr commented on core/state-machine/fuzz/src/lib.rs:
```diff
@@ -0,0 +1,254 @@
+	let mut len = 0;
+	let mut success = true;
+	for (key, value) in overlayed.iter_values(None) {
+
```
```suggestion
```

bkchr commented on core/state-machine/src/basic.rs:
```diff
@@ -205,13 +205,27 @@ impl Externalities for BasicExternalities {
 		warn!("Extensions are not supported by `BasicExternalities`.");
 		None
 	}
+
```
```suggestion
```

bkchr commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,515 @@
+	///
+	/// This method removes latest dropped values up to the latest valid value.
+	pub fn get_mut(&mut self, history: &[TransactionState]) -> Option<HistoriedValue<&mut V>> {
+
```
```suggestion
```

bkchr commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,515 @@
+}
+
+impl<V> History<V> {
+
```
```suggestion
```

bkchr commented on core/utils/historied-data/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "historied-data"
```
I'm not native speaker, but I think this should be renamed and also `substrate-` prefixed.

I think `substrate-historized-data` would be more correct.

Maybe @shawntabrizi can help us here? :)

shawntabrizi replied:
Neither sound right.

You want `historical-data`?

https://www.merriam-webster.com/dictionary/historical

cheme replied:
'historical' is sounding a bit funny to me (makes me think of chivalry and more generally historical events).

Maybe there is other possibilities :
here the purpose of the crate is to describe/manipulate a data that is stored with multiple representations depending on a given state (state being a point in history but could be something else for instance a blockchain fork plus a block number).

shawntabrizi replied:
`alternative-data`?

`parellel-data`?

I think `historical` is not really a word you should actively associate with physical history. When you clear your browser history, you would probably say you are clearing your historical data.

bkchr commented:
(Generally I'm okay with the direction of this pr, as long as we don't make anything slower^^)

bkchr commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,515 @@
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+type MemoryOnly<V> = smallvec::SmallVec<[HistoriedValue<V>; ALLOCATED_HISTORY]>;
```
Smallvec supports `no_std` as well.

bkchr commented on core/utils/historied-data/src/linear.rs:
```diff
@@ -0,0 +1,515 @@
+	}
+
+	#[cfg(any(test, feature = "test"))]
+	pub fn len(&self) -> usize {
```
Please remove the other implementation. Just make it public and good. It does not modify any state, so having it public will not harm anyone.

bkchr commented on core/utils/historied-data/src/lib.rs:
```diff
@@ -0,0 +1,84 @@
+	Pending,
+	/// Data is under change and can still be dropped.
+	/// This also mark the start of a transaction.
+	TxPending,
```
I think when we drop the prospective stuff from the overlayedchanges, we probably can merge `Pending` and `TxPending`?

bkchr commented on core/utils/historied-data/src/lib.rs:
```diff
@@ -0,0 +1,84 @@
+	/// Data is committed, but can still be dropped
+	/// using `discard_prospective` or `discard_transaction`
+	/// from a parent transaction state.
+	Prospective,
```
This probably can be removed as well.

cheme commented:
@bkchr , thanks for looking at it. Quick feedback:

#### Why do we still have prospective? Couldn't that be done with your feature as well?

Sergey had the same opinion, and I did start an implementation without it but rollback. Gist of it was:
- having a prospective state is a bit better for gc (but shouldn't really be a consideration).
- having prospective makes me feel better in the sense that it is a state that is not accessible from the runtime when other overlays are. But under the assumption that runtime are well written this is not a valid point.
- case like runtime panic are easier to reason about, but everything can be resolved to checking the height of the overlay to know if it is the first. But in this case we are back to the same logic, just not naming it explicitely (eg where we have a commit prospective here we will have commit all overlay but the n - 1, and if we want runtime to be safe we will check the overlay on commit and only commit if number is > 'starting ext eval height'): managing externally to the crate a prospective height counter.
But it is true that I have a use case where I clone the full overlay (when two type of eval are called) and with overlay only I would not (starting height would be 2 instead of 1).
So I should give it another try (and parameterized the gc with an equivalent to prospective height), with safeguard not in historied crate logic but outside.

#### I still don't understand why we have a garbage collector. I'm somehow allergic against garbage collectors :D Especially for such short lived stuff like the OverlayedChanges.

Garbage is mostly here because I did not feel good not having it, things did not feel complete, and did not want to argue about memory leakage. But I agree it should not really be call (trigger constant threshold values are set a bit high).

bkchr commented:
> @bkchr , thanks for looking at it. Quick feedback:
> #### Why do we still have prospective? Couldn't that be done with your feature as well?
> 
> Sergey had the same opinion, and I did start an implementation without it but rollback. Gist of it was:
> 
>     * having a prospective state is a bit better for gc (but shouldn't really be a consideration).
> 
>     * having prospective makes me feel better in the sense that it is a state that is not accessible from the runtime when other overlays are. But under the assumption that runtime are well written this is not a valid point.
> 
>     * case like runtime panic are easier to reason about, but everything can be resolved to checking the height of the overlay to know if it is the first. But in this case we are back to the same logic, just not naming it explicitely (eg where we have a commit prospective here we will have commit all overlay but the n - 1, and if we want runtime to be safe we will check the overlay on commit and only commit if number is > 'starting ext eval height'): managing externally to the crate a prospective height counter.
>       But it is true that I have a use case where I clone the full overlay (when two type of eval are called) and with overlay only I would not (starting height would be 2 instead of 1).
>       So I should give it another try (and parameterized the gc with an equivalent to prospective height), with safeguard not in historied crate logic but outside.
> 

I'm not saying we should not have the concept of prospective, we still need it, but it does not makes any sense to have 2 implementations for the same concept. As you said, you can just use your new feature to implement prospective, aka just call `start_transaction` or whatever it is called^^ I would probably go to a more generic naming `new_layer` or something similar.

> 
> #### I still don't understand why we have a garbage collector. I'm somehow allergic against garbage collectors :D Especially for such short lived stuff like the OverlayedChanges.
> 
> Garbage is mostly here because I did not feel good not having it, things did not feel complete, and did not want to argue about memory leakage. But I agree it should not really be call (trigger constant threshold values are set a bit high).

We probably can reduce the code by some lines when we drop the GC stuff? 

How does the GC works on a high level, what kind of information does it garbage collects?

cheme commented:
>  How does the GC works on a high level, what kind of information does it garbage collects?

GC goes through all overlay key once and for each key removes the values that are not needed in regard of current state:
- values of state index before committed (if we remove prospective/committed info  is just a start index), except last change (the committed reference value).
- Dropped overlay values. (in non full mode it will be only the terminal one to use only pop, but for non terminal instead of remove I should just clear the stored vec value instead of changing the value history layout (just realized)).

So hypothetical huge block this can

bkchr commented:
> > How does the GC works on a high level, what kind of information does it garbage collects?
> 
> GC goes through all overlay key once and for each key removes the values that are not needed in regard of current state:
> 
>     * values of state index before committed (if we remove prospective/committed info  is just a start index), except last change (the committed reference value).

This sounds like an operation that should be done on `commit`.

> 
>     * Dropped overlay values. (in non full mode it will be only the terminal one to use only pop, but for non terminal instead of remove I should just clear the stored vec value instead of changing the value history layout (just realized)).
> 

By drop you mean, I called `discard_transaction` (or whatever it is called) and still sounds like you should do the operation in place. Nevertheless, I would not expect to have so much discarded transactions. When we have one discarded transaction, we probably will end the entire enclosing transaction.

cheme commented:
> By drop you mean, I called discard_transaction (or whatever it is called) and still sounds like you should do the operation in place. Nevertheless, I would not expect to have so much discarded transactions.

'discard_transaction' is not doing anything except writing in the state +1 that there was a discard (same with start_transaction). This is interesting as you can manipulate transaction without cost, that way you can have some 'try' expression that user can use without thinking about performance.
Actually there is an additional cost but only when you write a new value (but it is easier to set a cost of writing a value that consider a possible append in a vec than setting a cost on a discard transaction that depends on the number of element and size of the state).
If we update values on discard the cost of iterating through all values will be (I think) too high: unless discard is seen as a rare operation, but it will probably be an easy dos vector then (or at least an operation that is difficult to price).

> When we have one discarded transaction, we probably will end the entire enclosing transaction.

Or branch to a different processing: those operations did fail or lead to a state that I do not want, rollback and try something else (that is basically the need that I understood: Ink have to manage its change for rollback).

> This sounds like an operation that should be done on commit.

It is indeed lighter than the gc, so it could be call here and there, depending on how much change a runtime is allowed to do in a block (whole struct lifetime is a block processing).
Iirc at some point I wanted to have this 'fast' (eager = false in fn gc) gc running more frequently, but end up finding it negligeable. I could still have a lower trigger for it on commit action.

But running it at every commit I will change the code to either:
- store keys to change values so I do not process the whole overlay (could be use for a precise cost model but is in itself costy).
- use stack of hashmap. Then the cost problem is on getting a value: it depends on the number of traversed map.

bkchr commented:
>  This is interesting as you can manipulate transaction without cost, that way you can have some 'try' expression that user can use without thinking about performance.

This is not correct, you will pay with higher memory usage. 

Nevertheless, why not overwrite the discarded value the next time we write to this value from another transaction? By doing this, we would not require to update the state on discard, but would also not require to run the garbage collection.

cheme commented:
> Nevertheless, why not overwrite the discarded value the next time we write to this value from another transaction?
> why not overwrite the discarded value the next time we write to this value from another transaction?

It is done, in 'get_mut' but I was thinking it does not cover all case, right now I think it does and cannot find which were those case (I remember at the time I was also thinking about how to implement per account transaction which I hope we will never implement).
I will try to go with it (alongside deletion of committed/prospective in the crate).

> This is not correct, you will pay with higher memory usage.
Yes, I said that afterward. But for sure current approach is not compatible with a cost model that refunds on discard transaction (or with an heuristic).

cheme commented:
I am putting back to A3-inprogress, since last changes if they looks overall good totally break the fuzzer with garbage collection, will look at it tomorrow.

cheme commented:
Putting back to reviewable.

So I change naming to historical, and apply deletion on 'get_mut' in case of two transaction getting merge (previously it only dropped discarded transaction).

To that end, there is a bit more operation going on resulting in a small drop in bench.

with this diff both version of this diff https://gist.github.com/cheme/1729ddb750b8fdda69f7012a8d5e1ec4,

I lose from updating committed values :
commit_drop_commit
    change: [-3.4183% -3.2588% -3.1007%] (p = 0.00 < 0.05)
commit_drop_commit_get
    change: [-2.3257% -2.2709% -2.2158%] (p = 0.00 < 0.05)

but between to refactored code and the fast patch version there is a slight around 1%
improvement(nothing really relevant).

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,577 @@
+	}
+
+	#[inline]
+	fn clear_terminal_values(
```
Yes it is here to avoid some code redundancy and is a bit artificial as a function, I will try to get something.

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,577 @@
+		self.history.len()
+	}
+
+	/// Get index of committed layer, this is
```
It does not feel really pertinent, I am removing it.

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,577 @@
+	}
+
+	/// Current number of inner states.
+	pub fn len(&self) -> usize {
```
Can do, it seems more related to the naming indeed (no simply a vec), so I guess it is better.

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,577 @@
+		let mut index = self.len();
+		if index == 0 {
+			return None;
+		}
```
It was to indicate that a 'states' shall always have at least a pending layer, but this comment do not make sense in this context since we do manage the 0 case, I am removing it (it was also somewhere else).

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,577 @@
+		self.clear_terminal_values(result, previous_switch, deleted)
+	}
+
+	#[inline]
```
That is probably not useful. I did add this function to avoid code redundancy but does not want a new function at all, so from bad habit I added it.
Last parameter is constant in one of its two calls, but lets trust the compiler.

bkchr commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+	fn mut_ref(&mut self, index: usize) -> &mut V {
+		&mut self.0[index].value
+	}
+
```
```suggestion
```

bkchr commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+
+}
+
+
```
```suggestion
```

bkchr commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+	// This is redundant with `history`, but is
+	// used to avoid some backwards iteration in
+	// `find_previous_tx_start` function.
+	previous_transaction: Vec<usize>,
```
I don't understand what this is doing.

cheme commented:
On the need of TxPending, there was a case scenario that I forget in last discussion when considering the cases of pruning.
with s being state and v being value:
s: 1pending
-> set v to 2
v: (1:2)
-> start transaction
s: 1:pending 2:txpending
-> set v to 3
v: (1:2) : (2:3)
-> start transaction
s: 1:pending 2:txpending 3: txpending
-> set v to 4
v: (1:2) : (2:3): (3:4)
-> commit transaction
s: 1:pending 2:txpending 3: pending
-> start transaction
s: 1:pending 2:txpending 3: pending 4:txpending
- set v to 4 : here before inserting 4 we drop (2,3) by knowing last tx
v: (1:2) : (3:4):(4:4)

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+		if index == 0 {
+			return None;
+		}
+		debug_assert!(states.len() >= index);
```
The check is only asserting that we are using the right state with the HistoryValue, this is really something that would be a bug and should not happen. So semantically it is an assertion, and it is unneeded so I debug assert.
I would rather remove it, I do not want to check that state use for query is the same as state use for update. This is basis assumption.

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+		self.0.truncate(index)
+	}
+
+	fn truncate_until(&mut self, index: usize) {
```
yes, truncate start is a bit odd, I will go with 'remove_until'.

cheme commented on core/utils/historical-data/src/linear.rs:
```diff
@@ -0,0 +1,575 @@
+// buffer specific functions.
+impl<V> History<V> {
+
+	fn get_state(&self, index: usize) -> HistoricalValue<&V> {
```
Yes, this was intentional (it replaces some [] access to the inner vec), but code can be change to use 'get'.
Generally private function here are mostly running over expectations of correct usage, I will try to add some panic comment in the doc, since there is may.

cheme commented:
Did a few change, I will look again at documentation (missing function and other) this afternoon.

cheme commented:
@bkchr , @pepyakin , I did implement the design change.
Things that I notice when implementing:
- the history simply store the index of the number of overlay use when stacking.
- the start number of transaction is one instead of zero. That way cleaning (discarding transaction for each value) do not have a specific condition (index must be reset after with 'finalize_discard').

Note for myself: ef0bcda is the last commit to contain previous implementation (one commit before is the last one with previous implementation use in other crates), it contains module layer explaining naming of the current historical data file.

cheme commented:
@pepyakin , thanks for the brief review, I addressed it in 19b1014 (giving you the reference to the commit because master rebase probably makes it difficult to check).

cheme commented on client/executor/src/deprecated_host_interface.rs:
```diff
@@ -1062,6 +1062,18 @@ impl_wasm_host_interface! {
 				}
 			})
 		}
+
```
Probably need to be in host_interface instead.

pepyakin commented:
@cheme so at the end, how the testing on Kusama went? Do we need another run? Anything stopping us from doing so?

@bkchr can you give this another look?

cheme commented:
Kusama testing was fine (I did synch up to head (around 780k at the time it was on trie 0.15 at the time)). Nothing prevent doing another run.

bkchr commented:
(Put a reminder on this for me. Sorry for the delay! :()

gavofyork commented:
Is this ready to merge? @bkchr @pepyakin you ok to sign off on it?

pepyakin commented:
The last time I looked, I haven't found big issues with it, although I'd like to give it another look. The fact that it syncs with Kusama and has fuzzing adds some more confidence. I see some problems wrt to docs, but I hope that we can iterate on in follow-ups.

cheme commented on utils/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,94 @@
+
+impl<V, I> History<V, I> {
+	/// Get current number of stored historical values.
+	pub fn len(&self) -> usize {
```
only fuzzer, I moved it back to being a test helpers method.

gavofyork commented on frame/support/src/storage/mod.rs:
```diff
@@ -25,6 +25,22 @@ pub mod hashed;
+/// Execute under a transactional layer.
+///
+/// If the result of execution is an error,
+/// the transactional layer get reverted; otherwhise
```
```suggestion
/// the transactional layer get reverted; otherwise
```

gavofyork commented on frame/support/src/storage/mod.rs:
```diff
@@ -25,6 +25,22 @@ pub mod hashed;
+pub fn with_transaction<R, E>(f: impl FnOnce() -> Result<R, E>) -> Result<R, E> {
+	sp_io::storage::start_transaction();
+	let r = f();
+	if r.is_ok() {
```
better to use a `match` here.

gavofyork commented on frame/support/src/storage/mod.rs:
```diff
@@ -25,6 +25,22 @@ pub mod hashed;
+/// it is committed.
+pub fn with_transaction<R, E>(f: impl FnOnce() -> Result<R, E>) -> Result<R, E> {
+	sp_io::storage::start_transaction();
+	let r = f();
```
@cheme what do you expect to gain by assigning from one single-letter-function call to a single-letter-variable?

```suggestion
	let result = f();
```

gavofyork commented on primitives/externalities/src/lib.rs:
```diff
@@ -156,6 +156,16 @@ pub trait Externalities: ExtensionStore {
+	fn storage_start_transaction(&mut self);
+
+	/// Discard a transactional layer, pending changes of every transaction behind this layer are
+	/// dropped (including committed changes) .
```
```suggestion
	/// dropped (including committed changes).
```

please pay attention to detail.

gavofyork commented on primitives/historical-data/README.md:
```diff
@@ -0,0 +1,18 @@
+
+This covers:
+- linear history driven data, eg. transactional layers for overlay.
+- long term storage with multiple branch, eg. offchain storage.
```
```suggestion
- long term storage with multiple branches, eg. off-chain storage.
```

gavofyork replied:
is that even the intended meaning?

cheme replied:
I am removing this line, it is something that is implemented in another PR, but since the code is currently not in the crate it does not make sense to keep it.

gavofyork commented on primitives/historical-data/README.md:
```diff
@@ -0,0 +1,18 @@
+- linear history driven data, eg. transactional layers for overlay.
+- long term storage with multiple branch, eg. offchain storage.
+
+General design is container where query and update requires global
```
```suggestion
General design is to be a container where query and update both require a global
```

i'm changing this from non-english to english. i've no idea if what i'm writing is what you were meaning @cheme .

gavofyork commented on primitives/historical-data/README.md:
```diff
@@ -0,0 +1,18 @@
+General design is container where query and update requires global
+history context.
+
+History is serialize as a per item basis.
```
Not english.

gavofyork commented on primitives/historical-data/README.md:
```diff
@@ -0,0 +1,18 @@
+
+History is serialize as a per item basis.
+
+This crates is `no_std` compatible as long as the `std` feature is not enabled.
```
```suggestion
This crate is `no_std` compatible as long as the `std` feature is not enabled.
```

gavofyork commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+}
+
+#[derive(Debug, PartialEq)]
+/// Results from cleaning a data with history.
```
```suggestion
/// The results from cleaning a data with history.
```

???

gavofyork replied:
No idea what you mean by "a data". This is not english.

gavofyork commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+/// It should be used to update from the calling context,
+/// for instance remove this data from a map if it was cleared.
+pub enum CleaningResult {
+	Unchanged,
```
Undocumented items.

gavofyork commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+
+#[derive(Debug, PartialEq)]
+/// Results from cleaning a data with history.
+/// It should be used to update from the calling context,
```
No idea what this is meant to mean.

gavofyork commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+	Cleared,
+}
+
+/// History of value and their state.
```
```suggestion
/// History of value and its state.
```

gavofyork replied:
Yeah, but what is the context in which is it used? This documentation is completely useless for a newcomer to the codebase.

gavofyork commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+}
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
```
No idea what you mean here @cheme

pepyakin commented:
I am not sure, are the concerns mentioned are fixed?

cheme commented:
@pepyakin , it should be ready for another round of review, even if I am not very confident with my documentation skills (I will take another look at it tomorrow).

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+}
+
+impl States {
+
```
```suggestion
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+
+	/// After a prospective was discarded, clear prospective history.
+	pub fn apply_discard_prospective<V>(value: &mut History<V>) -> CleaningResult {
+		if value.0.len() == 0 {
```
```suggestion
		if value.0.is_empty() {
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+			}
+			value.0.truncate(1);
+		} else {
+			value.0.clear();
```
Why don't we return `Cleared` here?

cheme replied:
:+1: that is an error.

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+	/// Update states when committing prospective.
+	/// A subsequent update of all related stored history is needed.
+	pub fn commit_prospective(&mut self) {
+		self.current_layer = 1;
```
That is the third function which does this.

I don't understand it.

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+	/// on all values from this prospective. It commits pending value and
+	/// clear existing history.
+	pub fn apply_commit_prospective<V>(value: &mut History<V>) -> CleaningResult {
+		if value.0.len() == 0 {
```
```suggestion
		if value.0.is_empty() {
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+	/// Discard a transactional layer.
+	/// A subsequent synchronous update of all related stored history is needed.
+	pub fn discard_transaction(&mut self) {
+		if self.current_layer > 0 {
```
```self.current_layer = self.current_layer.saturating_sub(1);```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+				index: State::Transaction(ix),
+			} = value.0[i] {
+				if ix > self.current_layer {
+					let _ = value.0.pop();
```
```suggestion
					value.0.pop();
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+				} else { break }
+			} else { break }
+		}
+		if value.0.len() == 0 {
```
```suggestion
		if value.0.is_empty() {
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+			} = value.0[i] {
+				if ix > self.current_layer {
+					let _ = value.0.pop();
+				} else { break }
```
Please put them onto new lines.

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+}
+
+impl<V> History<V> {
+
```
```suggestion
```

bkchr commented on primitives/historical-data/src/synch_linear_transaction.rs:
```diff
@@ -0,0 +1,320 @@
+	pub fn apply_commit_transaction<V>(&self, value: &mut History<V>) -> CleaningResult {
+		let mut new_value = None;
+		for i in (0 .. value.0.len()).rev() {
+			if let HistoricalEntry {
```
I don't understand what is going on here.

cheme commented on core/utils/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,120 @@
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Data storage containing multiple state for a value.
```
That is a bit odd, the change is on a file that does not exists anymore (I will look at the new location for it).

cheme commented on Cargo.toml:
```diff
@@ -133,6 +133,7 @@ members = [
 	"test/utils/transaction-factory",
 	"utils/build-script-utils",
 	"utils/fork-tree",
+	"utils/historical-data",
```
was done in https://github.com/paritytech/substrate/pull/3263/commits/0f8bee672d1e1d8d3adb7613592e78dccb8b5f8f .

cheme commented on primitives/historical-data/src/lib.rs:
```diff
@@ -0,0 +1,105 @@
+	/// No inner data was changed, even technical
+	/// data, therefore no update is needed.
+	Unchanged,
+	/// Any data got modified, therefore an
```
I will remove the update sentence. It was in case the in memory data needs to be serialized, but it is rather implicit.

cheme commented:
@bkchr , thanks for the review, I think I went over most of the item. Documentation for non-understandable code might not be about .
About duplicate code, I kept the multiple `self.current_layer = 1` functions, because they correspond to different state operation that would be different implementation of a global state trait, the fact that implementations are the same is because we do not support asynch changes. Still I can rename them to `revert_to_initial_state` if needed.

cheme commented:
I did remove historical crate, code get flattened into state-machine. It is probably not a issue since #3774 is closed, and it will likely get redesign differently (thinking of a clean separation with existing code so the historical use do not need to merge within substrate but just is a specific implementation detail).

gavofyork commented on primitives/state-machine/src/overlayed_changes.rs:
```diff
@@ -625,21 +803,41 @@ impl OverlayedChanges {
+/// This could also be easilly replace if btreemap gets
+/// an implementation for retain in the future.
+fn retain<K: Ord + Clone, V, F>(map: &mut BTreeMap<K, V>, mut f: F) where
+    F: FnMut(&K, &mut V) -> bool, {
```
spaces!!

gavofyork replied:
`{` belongs on own line

gavofyork commented on primitives/state-machine/src/transaction_layers.rs:
```diff
@@ -0,0 +1,506 @@
+/// transaction operation.
+#[derive(Debug, PartialEq)]
+pub(crate) enum LayeredOpsResult {
+	/// No inner data was changed, even technical
```
what is "technical data"?

gavofyork replied:
perhaps you mean "metadata" or "housekeeping data"?

gavofyork commented on primitives/state-machine/src/transaction_layers.rs:
```diff
@@ -0,0 +1,506 @@
+
+	/// Set a new value, this function expect that
+	/// `number_transactions` is a valid state (can fail
+	/// otherwhise).
```
```suggestion
	/// otherwise).
```

gavofyork commented:
This will need extensive documentation on the complexity implications before it can be used in runtimes. it looks like it introduces an `O(N)` cost to all transaction layer operations where `N` is the total number of storage changes throughout all previous transaction layers. This means it can't be easily utilised by the contracts module, since it would be possible to make a quadratic attack (paying gas for a transaction that changes 1000s of storage items, and then making many trivial (and thus cheap) calls which require iterating through 1000s of not-yet-committed items).

cheme commented:
Yes I think it is very problematic (putting pr in status 'got issue'), even more when you consider that N is not only change done by a single extrinsic but also all previous extrinsic in block.
I see three ways of fixing it:
- Revert to first implementation where transaction operations were zero cost: we simply keep an history of change that is later used when accessing data ( data change history cleaning being done on access or mutable access or on garbage collection but the garbage collection should be removed from the cost here : when storing/changing value we pay for the memory that is allocated and gc is an optimization for extreme case). In this case the cost of freeing old values is included into the cost of change/creation but only applied on access (actually the fact that it is on mutable access breaks a bit this assumption, inner mutability would be needed to make things totally bounded).
- Do some bookkeeping of updated keys per transaction layer so on transaction operations we only modify the changed keys, and cost got include in change/modify operation (appending key to vec and doing a clean or commit of latest value to this vec), but it is bounded.
- Use one hashmap per transaction layer. Here the cost of removing/committing a layer is dependent on the number of value in the layer, but it is not such an issue because that is only something that is local to a single extrinsic processing. So for transaction operation cost this is  less bad: the 1000s changes are payed by the same user and cost only apply once per transaction.
The main issue here is the cost does grow for key value query over the number of transaction layer so it is a similar issue.

CC/ @bkchr , @pepyakin, my preference is obviously my first implementation, but we revert from it for some reason so please tell me what you think is best.

gavofyork commented:
So the complexity profile we're looking for it:
- Read storage: O(1)
- Write storage: O(1)
- Create new transaction: O(1)
- Rollback transaction: O(N) (N = number of storage writes since creation, including for nested transactions)
- Commit transaction: O(N)

In none of these is nesting-depth, nor pending storage writes allowed to have an effect on the time for a read/write.

cheme commented:
With:
- NO number of nested transaction
- NB number of storage write since block evaluation start
- N number of storage change since last Transaction (including nested transaction).

Current implementation is :
    - Read storage: O(1)
    - Write storage: O(1)
    - Create new transaction: O(1)
    - Rollback transaction: O(NB)
    - Commit transaction: O(NB)
First solution (previous implementation):
    - Read storage: O(~1)
    - Write storage: O(~1)
    - Create new transaction: O(1)
    - Rollback transaction: O(1)
    - Commit transaction: O(1)
Second solution (key change book keeping):
    - Read storage: O(1)
    - Write storage: O(1)
    - Create new transaction: O(1)
    - Rollback transaction: O(N)
    - Commit transaction: O(N)
Third solution (stacked hashmaps):
    - Read storage: O(NO)
    - Write storage: O(1)
    - Create new transaction: O(1)
    - Rollback transaction: O(N) (if we consider disallocation cost)
    - Commit transaction: O(N) (but 2* map access)

So the second solution may be the one that match (I find the first way better but code complexity was a no go previously).

DemiMarie-parity commented:
> With:
> 
>     * NO number of nested transaction
> 
>     * NB number of storage write since block evaluation start
> 
>     * N number of storage change since last Transaction (including nested transaction).
> 
> 
> Current implementation is :
> - Read storage: O(1)
> - Write storage: O(1)
> - Create new transaction: O(1)
> - Rollback transaction: O(NB)
> - Commit transaction: O(NB)
>
> First solution (previous implementation):
>
> - Read storage: O(~1)
> - Write storage: O(~1)
> - Create new transaction: O(1)
> - Rollback transaction: O(1)
> - Commit transaction: O(1)
>
> Second solution (key change book keeping):
>
> - Read storage: O(1)
> - Write storage: O(1)
> - Create new transaction: O(1)
> - Rollback transaction: O(N)
> - Commit transaction: O(N)
> 
> Third solution (stacked hashmaps):
> 
> - Read storage: O(NO)
> - Write storage: O(1)
> - Create new transaction: O(1)
> - Rollback transaction: O(N) (if we consider disallocation cost)
> - Commit transaction: O(N) (but 2* map access)
> 
> So the second solution may be the one that match (I find the first way better but code complexity was a no go previously).

Would the first solution be strictly O(1) for reads and writes if interior mutability was used?  Would it be possible to make all accesses require `&mut self`?

cheme commented:
> Would the first solution be strictly O(1) for reads and writes if interior mutability was used?

Currently there could be NO drop of values but it occurs only once (first solution is basically making the transaction change from second solution lazy).

Indeed without inner mutability we can have some NO additional access to pointers (NO Vec access to value change and corresponding status in history (direct pointer access too)) for 'get' operation.
This only happen when multiple layers are discarded and we access the value afterward (on commit the value is ok, we just do not reclaim some unused space). But it is rather easy to do in a single extrinsic call.

So, with inner mutability we can still have NO dropped transaction to clean on access (but it is not a big operation: check NO values and possibly drop them (in fact the very first implementation did only update latest value like described here, but at some point to simplify the gc it was switched to a more impacting cleaning but honestly this cleaning operation if done once is really small)).

> Would it be possible to make all accesses require &mut self?

That is probably the cleanest way to do it (inner mutability will probably lead to some awkward Mutex because of some traits being Sync), but it changes the api significantly for some technical implementation details.

But currently the first solution is not wanted (we switch to current implementation because code complexity was not considered worth the gain :( ).

cheme commented:
I did implement the second solution on this branch https://github.com/cheme/substrate/compare/overlay_change...cheme:overlay_change_rc , it makes things a bit more complicated (it could be good to check perf diff if storing directly key instead of weak rc).
The general perf likely drops a lot (on existing test between 30 to 40 % but it is does not realy mean anything).

Choice to switch to selective cleaning is currently set by those measurement, and the
redundancy of key change between layer needs an heuristic or simply run calculation over
first layer (for setting heuristic value we need to ensure that setting n time the same value in n layer always 
a commit :
- 100 keyval: 
	- 1/2 : full better by 8%
	- 1/4 : selective better by 4%
- 1000 keyval:
	- 1/4 : full better by 11%
	- 1/8: selective better by 1%
- 10_000 keyval:
	- 1/8: full better by 10%
	- 1/16: selective better by 1%
Keeping it as a separate branch for now.

gavofyork commented:
When it's O(~1) for the first solution, what does the "~" imply? Is it sometimes O(N)?

cheme commented:
It can go up to O(NO) with NO the number of consecutive discarded transaction where a write for this accessed key occurs:
- when reading we have to skip those discarded transaction when reading (NO comparison of index in a vec).
- when writing we drop those value

This reverts to O(1) at first writing (or reading if we use inner mutability), so it is something not very relevant.
For a cost model this operation could be seen a part of the writing cost of a value (it is strictly right only if we do update on read but it that read could mutate the overlay).

gavofyork commented:
so someone could attack this by creating 10,000 discarded transactions that write to a single value and then read from that value 10,000 times. this would result in 100,000,000 operations but a cost of just `10,000w + 10,000r`.

this is not O(1), therefore it cannot be used as it stands.

> For a cost model this operation could be seen a part of the writing cost of a value

no, it could not, since the cost is paid for every read.

cheme commented:
Agree this is not O(1) without implementing mutability on read access.
Conclusion, if we switch back to previous implementation (I would need consent from @bkchr and @pepyakin), it will need to implement the cleanup of value on first read (a little bit involved).
If we stay with current implementation we need bookkeeping of change per layer as implemented in branch yesterday (a bit costly, it will probably degrade current perf a bit even for the case where no transaction are used).
Stacking hashmap is not fine to me due to O(N) in read.

gavofyork commented:
I guess "bookkeeping of change per layer as implemented in branch yesterday" is the second solution?

i would probably attempt to implement both and benchmark. certainly having every operation be `O(1)` is nice, but i guess as long as there are no hidden `N**2`s in there then it's ok.

cheme commented:
Yes that was the second solution.
First solution was trading computation cost of the O(N) transactional operation of the second solution against some additional memory consumption (operation being lazy), so basic benches were at the time in its favor. With additional bookkeeping for the second solution this will probably move a bit more toward the first solution.
The decision to switch from the first solution to the current implementation was taken at the time to simplify the code, considering that the perfs here were not a bottleneck.
But with the additional bookkeeping code complexity, this argument may not hold anymore.

DemiMarie-parity commented:
@cheme @gavofyork How difficult would it be to remove the need to be `Sync`?  That would allow the use of interior mutability without synchronization.

Otherwise, I am worried we will explode in lock contention, unless we use lock-free data structures.  I know I don’t want to try to debug those.

cheme commented:
Found back this: https://github.com/paritytech/substrate/pull/3263#issuecomment-532612805 , basically the iterator over a refcell was painful to implement.

I find the idea of using '&mut self' on 'get' operation more promising (the overlay during block processing is owned by a single thread), IIRC there may be some trait that access the storage in a readonly manner, for those it may be a problem, but at the end those trait do not need to clean since they are not running over modification, so I have good hope there may be a simple solution.

But yes, in the hypothesis of a RWLock over a global read state for runtime workers, just changing 'get' operation to use '&mut' will add lock contention : I can think of running 'get_and_does_it_need_clean(&self,...' and call a clean operation (with lock) only if needed: we double the job but it stays constant time and remove most of the additional lock contention.

DemiMarie-parity commented:
> Found back this: [#3263 (comment)](https://github.com/paritytech/substrate/pull/3263#issuecomment-532612805) , basically the iterator over a refcell was painful to implement.
> 
> I find the idea of using '&mut self' on 'get' operation more promising (the overlay during block processing is owned by a single thread), IIRC there may be some trait that access the storage in a readonly manner, for those it may be a problem, but at the end those trait do not need to clean since they are not running over modification, so I have good hope there may be a simple solution.

What about having a clean after block processing is complete?  The idea is that you get an RAII guard that gives you mutable access; once this is dropped, everything is cleaned.

Why does a get need to cause cleaning?  Could we do cleaning during transaction end or start?

> But yes, in the hypothesis of a RWLock over a global read state for runtime workers, just changing 'get' operation to use '&mut' will add lock contention : I can think of running 'get_and_does_it_need_clean(&self,...' and call a clean operation (with lock) only if needed: we double the job but it stays constant time and remove most of the additional lock contention.

Not really.  Most RWLocks contend even for reads, even though readers do not conflict with each other.  Avoiding that requires something like a seqlock or RCU, where taking a read lock does not require any writes.  RCU is fiendishly complicated, however, and seqlocks require `Copy` data.

If we can require all accesses to take `&mut self`, we can make things much cleaner.  One option is to

cheme commented:
> What about having a clean after block processing is complete? The idea is that you get an RAII guard that gives you mutable access; once this is dropped, everything is cleaned.

Those values get dropped at the end of block processing. We iterate on all of them and send only the latest pending value to 'client', so this is not needed.
The actual cleaning is just freeing the tail of the vec containing the changes, and the actual operation that goes O(N)  in the number of discarded transactions is a iteration over the N last element of the vec with each time a comparison with another index (plus pointer arithmetic to this other vec), so it is really not something heavy.

> Why does a get need to cause cleaning?

First 'get' or 'get_mut' need cleaning, otherwise all subsequent 'get' call will need to iterate over all uncleaned discarded transaction making the query more costly depending on the transaction context. By cleaning on first access, we ensure all following accesses are O(1).

> Could we do cleaning during transaction end or start?

In the second solution (and in current code) we run the cleaning at every transaction end. There is an issue with it in current code because we do the cleaning on every key, so it depends on the number of key. So in the branch with bookkeeping the cleaning is only done on the keys that get modified for the discarded transactions.

> Not really. Most RWLocks contend even for reads, even though readers do not conflict with each other. 

So, having more read access than write is a good thing, that is what I meant. (it does not reduce the number of access to the underlying atomic integer indeed).
Not sure about seqlock or RCU.

cheme commented:
So I do have three different branch, for three different solutions:
- https://github.com/cheme/substrate/tree/overlay_change_lazy_mut : that is the first solution with some function switched to mutable.
This change some runtime function to mutable, without testing I think it is not breaking api (wasm don't care on mutable).
At first it felt really wrong to me because it changes mutability of some api for strictly technical reason, but after some more thinking it feels not that bad (maybe having mutable primitive could even be something needed for similar reason in runtime worker design).
- https://github.com/cheme/substrate/tree/overlay_change_lazy_inner : that is the first solution but using inner mutability (refcell or rwlock for test). It was smoother than expected to put in place (actually some client calls where already using refcell). The branch got a few thing to refactor but it seems rather ok. My main concern here is that it is not a strictly needed inner mutability and that it can lead to new issue (mostly double mut access on recursive call, I did spot one in next_key function but there may remain some).
- https://github.com/cheme/substrate/tree/overlay_change_rc : that is the second solution (current code with bookkeeping of change keys), it uses some weakrc to avoid doing some uniqueness check, but that is only a proposal. Existing bench are pretty bad when comparing with the 2 others branch, but that is mainly related to bookkeeping, that should not be a main consideration. 

My preference goes toward the first branch, but I am not taking the decision here (we switch off from this solution previously so I prefer to wait on a consensus before pushing the last effort to document, test in polkadot, clean, run more tests, rework documentation of the chosen solution).

gavofyork commented:
https://github.com/cheme/substrate/tree/overlay_change_lazy_mut seems like the sensible path to take at this stage.

cheme commented:
I did include lazy_mut, also run a polkadot (https://github.com/cheme/polkadot-1/tree/overlay_change_nb_2) sync for kusama last night. Putting the PR back to 'please review' state.

cheme commented on primitives/state-machine/src/ext.rs:
```diff
@@ -333,48 +332,54 @@ where
+				(_, Some(overlay_key)) => if overlay_key.1.value.is_some() {
+					return Some(overlay_key.0.to_vec())
+				} else {
+					// TODO make this function non recursive to avoid this clone
```
No those TODO were just a note to myself, I am removing them.

cheme commented:
This is stale, closing it, @athei will be working on the subject, keeping it open will only create confusion.
Still if some want to use it please don't hesitate to contact me (it got nice complexity property).
